File: cpp/graph/dfs.cc
#include <deque>
#include <unordered_set>
#include <vector>

using namespace std;

class Node {
 public:
  Node(int _i) : i(_i) {}
  void AddAdj(int v) {
    adj.push_back(v);
  }
  int i;
  vector<int> adj;
};

int DFS(int N, const vector<Node> &graph, int start, int find) {
  vector<bool> seen(N, false);
  deque<int> q;
  q.push_front(start);
  seen[start] = true;
  while (!q.empty()) {
    int u = q.front();
    q.pop_front();
    if (graph[u].i == find) return u;
    for (auto &v : graph[u].adj) {
      if (!seen[v]) {
        seen[v] = true;
        q.push_front(v);
      }
    }
  }
  return -1;
}

File: python/graph/dfs.py
from collections import deque

class Node:
    def __init__(self, i):
        self.i = i
        self.adj = []

    def add_adj(self, v):
        self.adj.append(v)


def DFS(start, find):
    q = deque([start])
    seen = set([start])
    while q:
        u = q.pop()
        if u.i == find:
            return u
        for v in u.adj:
            if v not in seen:
                seen.add(v)
                q.append(v)
    return None

